pkgs/top-level/default.nix

```diff
@@ -57,24 +56,36 @@ let
   platform = if platform_ != null then platform_
     else config.platform or platformAuto;
 
-  topLevelArguments = {
-    inherit system bootStdenv noSysDirs config crossSystem platform lib;
-  };
+  # A few packages make a new package set to draw their dependencies from.
+  # (Currently to get a cross tool chain, or forced-i686 package.) Rather than
+  # give `all-packages.nix` all the arguments to this function, even ones that
+  # don't concern it, we give it this function to "re-call" nixpkgs, inheriting
+  # whatever arguments it doesn't explicitly provide. This way,
+  # `all-packages.nix` doesn't know more than it needs too.
+  #
+  # It's OK that `args` doesn't include the defaults: they'll be
+  # deterministically inferred the same way.
+  mkPackages = newArgs: import ./. (args // newArgs);
 
   stdenvAdapters = self: super:
     let res = import ../stdenv/adapters.nix self; in res // {
       stdenvAdapters = res;
     };
 
   trivialBuilders = self: super:
-    (import ../build-support/trivial-builders.nix {
+    import ../build-support/trivial-builders.nix {
       inherit lib; inherit (self) stdenv stdenvNoCC; inherit (self.xorg) lndir;
-    });
+    };
 
-  stdenvDefault = self: super: (import ./stdenv.nix topLevelArguments) {} pkgs;
```

Comment:
Yeah, in my pull request (#18386), in `top-level/stdenv.nix`, if crossSystem is not null, then I make the final the stdenv using the default stdenv like this:

   (mkPackages { bootStdenv = defaultStdenv; }).stdenvCross

So I think that's what @Ericson2314 means when he says there will be another level of bootstrapping.  So ultimately stdenv.nix does not need to use `pkgs`, but it will need `mkPackages`.